pkg/controller/gcppubsub/reconcile.go

```diff
@@ -0,0 +1,326 @@
+		return src, err
+	}
+	src.Status.MarkSink(sinkURI)
+
+	sub, err := r.createSubscription(ctx, src)
+	if err != nil {
+		logger.Error("Unable to create the subscription", zap.Error(err))
+		return src, err
+	}
+	src.Status.MarkSubscribed()
+
+	_, err = r.createReceiveAdapter(ctx, src, sub.ID(), sinkURI)
+	if err != nil {
+		logger.Error("Unable to create the receive adapter", zap.Error(err))
+		return src, err
+	}
+	src.Status.MarkDeployed()
+
+	return src, nil
+}
+
+func (r *reconciler) addFinalizer(s *v1alpha1.GcpPubSubSource) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Insert(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func (r *reconciler) removeFinalizer(s *v1alpha1.GcpPubSubSource) {
+	finalizers := sets.NewString(s.Finalizers...)
+	finalizers.Delete(finalizerName)
+	s.Finalizers = finalizers.List()
+}
+
+func (r *reconciler) createReceiveAdapter(ctx context.Context, src *v1alpha1.GcpPubSubSource, subscriptionId, sinkURI string) (*v1.Deployment, error) {
+	ra, err := r.getReceiveAdapter(ctx, src)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logging.FromContext(ctx).Error("Unable to get an existing receive adapter", zap.Error(err))
+		return nil, err
+	}
+	if ra != nil {
+		logging.FromContext(ctx).Desugar().Info("Reusing existing receive adapter", zap.Any("receiveAdapter", ra))
+		return ra, nil
+	}
+	svc, err := r.makeReceiveAdapter(src, subscriptionId, sinkURI)
+	if err != nil {
+		return nil, err
+	}
+	err = r.client.Create(ctx, svc)
+	logging.FromContext(ctx).Desugar().Info("Receive Adapter created.", zap.Error(err), zap.Any("receiveAdapter", svc))
+	return svc, err
+}
+
+func (r *reconciler) getReceiveAdapter(ctx context.Context, src *v1alpha1.GcpPubSubSource) (*v1.Deployment, error) {
+	dl := &v1.DeploymentList{}
+	err := r.client.List(ctx, &client.ListOptions{
+		Namespace:     src.Namespace,
+		LabelSelector: r.getLabelSelector(src),
+		// TODO this is only needed by the fake client. Real K8s does not need it. Remove it once
+		// the fake is fixed.
+		Raw: &metav1.ListOptions{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: v1.SchemeGroupVersion.String(),
+				Kind:       "Deployment",
+			},
+		},
+	},
+		dl)
+
+	if err != nil {
+		logging.FromContext(ctx).Desugar().Error("Unable to list deployments: %v", zap.Error(err))
+		return nil, err
+	}
+	for _, dep := range dl.Items {
+		if metav1.IsControlledBy(&dep, src) {
+			return &dep, nil
+		}
+	}
+	return nil, apierrors.NewNotFound(schema.GroupResource{}, "")
+}
+
+func (r *reconciler) getLabelSelector(src *v1alpha1.GcpPubSubSource) labels.Selector {
+	return labels.SelectorFromSet(getLabels(src))
+}
+
+func getLabels(src *v1alpha1.GcpPubSubSource) map[string]string {
+	return map[string]string{
+		"knative-eventing-source":      controllerAgentName,
+		"knative-eventing-source-name": src.Name,
+	}
+}
+
+func (r *reconciler) makeReceiveAdapter(src *v1alpha1.GcpPubSubSource, subscriptionId, sinkURI string) (*v1.Deployment, error) {
```

Comment:
please move this to a resources dir for the controller

Reply:
Done.